lib/knetwork/kdns/records.go

```diff
@@ -0,0 +1,203 @@
+// are of the currently support record type. If any errors occur, it will write to the general error channel
+func (rc *RecordController) AddRecords(rr []dns.RR) {
+	for _, r := range rr {
+		switch r.Header().Rrtype {
+		case dns.TypeA:
+			rc.aRecordOps <- recordOperation{Type: RecordWrite, Record: r}
+		case dns.TypeTXT:
+			rc.txtRecordOps <- recordOperation{Type: RecordWrite, Record: r}
+		default:
+			rc.ErrorChan <- fmt.Errorf("kdns currently does not support record type %s: full record: %v",
+				r.Header().String(), r.String())
+		}
+	}
+}
+
+// SetRecords will hard replace records of type recordType in controller.
+// are of the currently support record type. If any errors occur, it will write to the general error channel
+func (rc *RecordController) SetRecords(rr []dns.RR) {
+	var txtRecords []dns.RR
+	var aRecords []dns.RR
+	for _, r := range rr {
+		switch r.Header().Rrtype {
+		case dns.TypeA:
+			aRecords = append(aRecords, r)
+		case dns.TypeTXT:
+			txtRecords = append(txtRecords, r)
+		default:
+			rc.ErrorChan <- fmt.Errorf("kdns currently does not support record type %s: full record: %v",
+				r.Header().String(), r.String())
+		}
+	}
+	if len(txtRecords) != 0 {
+		rc.txtRecordOps <- recordOperation{Type: RecordWriteForce, Records: txtRecords}
+	}
+	if len(aRecords) != 0 {
+		rc.aRecordOps <- recordOperation{Type: RecordWriteForce, Records: aRecords}
+	}
+}
+
+// EditRecords will replace records that match the keywords. if the supplied records also match
+// are of the currently support record type. If any errors occur, it will write to the general error channel
+func (rc *RecordController) EditRecords(rrs []dns.RR, keywords []string) {
+	var txtRecords []dns.RR
+	var aRecords []dns.RR
+	for _, r := range rrs {
+		switch r.Header().Rrtype {
+		case dns.TypeA:
+			aRecords = append(aRecords, r)
+		case dns.TypeTXT:
+			txtRecords = append(txtRecords, r)
+		default:
+			rc.ErrorChan <- fmt.Errorf("%w: %v", UnSupportedTypeErr, r.Header().Rrtype)
+		}
+	}
+	if len(aRecords) != 0 {
+		rc.aRecordOps <- recordOperation{Type: RecordEdit, Records: aRecords, Keywords: keywords}
+	}
+	if len(txtRecords) != 0 {
+		rc.txtRecordOps <- recordOperation{Type: RecordEdit, Records: txtRecords, Keywords: keywords}
+	}
+}
+
+// DeleteRecords will delete records that match the keywords provided. Case sensitive/no processing is done
+// TODO(adam): support regex?
+func (rc *RecordController) DeleteRecords(keywords []string, recordType uint16) {
+	switch recordType {
+	case dns.TypeA:
+		rc.aRecordOps <- recordOperation{Type: RecordDelete, Keywords: keywords}
+	case dns.TypeTXT:
+		rc.txtRecordOps <- recordOperation{Type: RecordDelete, Keywords: keywords}
+	default:
+		rc.ErrorChan <- fmt.Errorf("%w: %v", UnSupportedTypeErr, recordType)
+	}
+}
+
+// NewRecordController create a new controller for a specified origin, or basename;
+func NewRecordController() *RecordController {
```

Comment:
ok, let me see if I understand the design correctly:
1) for each name, eg, `worker-01.corp.enfabrica.net` there is a RecordController
2) a record controller holds two go coroutines, one for A records, one for TXT records.
3) each goroutine has a private list of records
4) the goroutine keeps waiting for: a) operations to edit the set of records, b) requests to read the records.
5) when records are edited (a), they are changed in the same thread. When they are read, they are sent over the channel.

What worries me here is 1) the leaking of goroutines, as records are added and deleted, 2) the # of goroutines, eg, if we have 100 names, there's going to be ... 3x goroutines? 3) an operation like a simple enumeration of the hosts (for a web interface? or to add prometheus monitoring?) will require querying all names? waking them all. 4) the moment we address those issues (eg, reclaiming goroutines, copying the array passed around - or having locks around those arrays) I think this code will be much more complex (cleanup is hard to do without deadlocks and weird race conditions).

I'm impressed by this design, truly - but in terms of complexity/debugging/concurrency issues, your original design using a simple map with locking was significantly simpler.

If I had to guess, at most will have 10s of queries per second reading, at most we'll have a couple changes per day, we will have however O(100s) of different names.  I would personally favor reducing complexity and lines of code over performance here

At this level, an `RWMutex` should be ok? A `sync.Map` would introduce some annoying type casting, but would probably lead to the same level of scalability than you have here? Getting rid of a global mutex, while still being very simple. We could scale up on goroutines in the DNS listener (or is it using a pool already?).

Reply:
Yes you understand the design correctly. 
And going over the concern in order:
1) They don't leak, see the added tests. All the transient channels are buffered, and the unbuffered ones get cleaned up, 

2) Correct. I'm ok with this.

3. Because of situations like this is why I rewrote it for nonblocking records. The scraping done by external parties is arbitrary, ttl is user configurable and the records are beings used in other systems. Any sort of deadlock would screw up multiple systems. With a mutex and a runaway script it would be easy to congest the dns.

4. I'm not sure if you caught on, but the pure read time without any operations is actually constant. Even a sync.Map (which I looked into because it does offer nicer sharding, which is nice and does scale better for higher read scenarios) begins significantly slowing down at 14000 records roughly. Which might seem like a lot until it isn't. 

4 cont. I won't argue it's simpler, but I will argue that it's tested. And goroutine ownership isn't exactly out of left field. If this gets merged I might take the first half of your comment and make it the readme even. I generally don't like speculating as to how much people will use X service because smart people tend to be very creative (and not cache friendly).

Reply:
Did you measure the performance? In 4, even if the code looks constant, it won't be: a simple enumeration will cause goroutines to be woken up and put to sleep constantly, while reading and writing into a channel. Waking and putting goroutines to sleep based on channel reading/writing ... I expect it to be significantly more costly than `sync.Map` mutexes - consider that coroutine management and channel still require some form of locking, atomics, and/or preempting.

If I run this microbenchmark here: https://gist.github.com/craigmj/5770480, similar to your code? It uses a channel to perform map operations with a serving goroutine, 10 goroutines on 8 CPU:
```
carlo@penguin:~/test$ ./test 
In parallel on 8 CPUs with 10 goroutines
GoMap:       1.356325302s
GoMap1Chan:  1.283012764s
SyncMap:     410.67364ms
```

But even this does not really matter? If we ever get to 14k records, we'll have a full team of developers, and we'll likely have other problems than this (mutliple server instances? monitoring? sharding? ...).

IMHO, what can bite us more than anything else are bugs like leaks of goroutine and concurrency, which there were a couple of in this PR. Also, ability to iterate over the code, adding functionalities as we learn what we need exactly. My recommendation would to focus on a) simplicity, b) defensiveness/reliability - more than performance.

Given the investment in time work, I'm ok moving forward with this once we address the pending comments, but I would argue that as we add features - or find other bugs - we should strive to simplify this code.

Some references:
https://medium.com/@deckarep/the-new-kid-in-town-gos-sync-map-de24a6bf7c2c - shows that a simple map with RWMutex can take about ~20 Million lookups per second, does not change with core numbers (scales very poorly). sync.Map can take about 4 Million lookups per second, but scales linearly with cores (eg, at about 4-5 cores, beats a normal map).